Make dynamic build use system librdkafka headers

KJTsanaktsidis opened the pull request:
cgo adds -I${SRCDIR} to the header search path by default; therefore,
using #include <librdkafka/rdkafka.h> can actually still find the
bundled librdkafka header in preference to the system one even when
`-tags=dynamic` is specified.

This change instead makes the static build set a preprocessor macro
USE_VENDORED_LIBRDKAFKA, which is used in a new header rdkafka_select.h
to either import the bundled one (with a special name, and quotes), or
the system one (with the <> import and the expected name).

Fixes #514

KJTsanaktsidis commented:
I couldn't actually _run_ the `./bundle-import.sh` script to check that my changes to it work, because I don't know where to find the `librdkafka-static-bundle-<VERSION>.tgz` artifact, and the nuget script in librdkafka can't build it without access to the `librdkafka-ci-packages` S3 bucket.

KJTsanaktsidis commented:
Think the go1.9 failures are the existing ones, but rewriting the imports in the examples seems to get the build to work elsewhere.

KJTsanaktsidis commented:
[clabot:check]

KJTsanaktsidis commented:
Well it only took nearly two months, but we (Zendesk) finally signed the corporate CLA 🎉 

The failing go 1.9 build looks to be what's fixed by https://github.com/confluentinc/confluent-kafka-go/pull/549

@edenhill - would ❤️ for you to take a look at this when you get a moment. Thanks!

edenhill commented on examples/admin_create_topic/admin_create_topic.go:
```diff
@@ -20,7 +20,7 @@ package main
 	"context"
 	"fmt"
-	"gopkg.in/confluentinc/confluent-kafka-go.v1/kafka"
+	"github.com/confluentinc/confluent-kafka-go/kafka"
```
These import changes are not related to this fix, please revert.

edenhill commented on kafka/00version.go:
```diff
@@ -21,7 +21,7 @@ import (
 
 /*
-#include <librdkafka/rdkafka.h>
+#include "rdkafka_select.h"
```
The "rdkafka_" prefix is used by librdkafka, so to be future-proof it is better if we use something else in the Go client, maybe "select_rdkafka.h".

edenhill commented on kafka/generated_errors.go:
```diff
@@ -1,9 +1,9 @@
 package kafka
 // Copyright 2016-2020 Confluent Inc.
-// AUTOMATICALLY GENERATED ON 2020-04-08 13:17:23.310983 +0200 CEST m=+0.001902312 USING librdkafka 1.4.0
+// AUTOMATICALLY GENERATED ON 2020-09-01 12:11:14.085648 +1000 AEST m=+0.000449739 USING librdkafka 1.4.2
```
Rebase on latest master, we're now importiung v1.5.2

edenhill reviewed (changes requested):
Great stuff!
Some things to fix:
 - Rebase on latest master
 - Rename to select_rdkafka.h to avoid future collisions
 - Remove the unrelated import changes
 - Win!

KJTsanaktsidis commented:
@edenhill thanks for the review. Are you able to point me to where I can get a copy of librdkafka-static-bundle-v1.5.2.tgz from? As far as I can see, the travis builds are uploaded to a private S3 bucket and so I can't make this archive myself using the `./release.py --class StaticPackage` script in librdkafka.

It would be good to have a copy of that so I can confirm that my changes to the import script work as expected.

edenhill commented:
[librdkafka-static-bundle-v1.5.2.tgz.zip](https://github.com/confluentinc/confluent-kafka-go/files/5535476/librdkafka-static-bundle-v1.5.2.tgz.zip)
Unzip this file (github will not allow .tgz uploads..)

KJTsanaktsidis commented:
@edenhill Rebased and addressed most of your comments, PTAL! Two things outstanding:

Firstly - not sure how you want to handle git history around here - I was tempted to squash the re-import of the librdkafka bundle (which regenerated the build_*.go files and a couple other things) into "make dynamic build use system librdkafka headers". If you want me to squash it down, I can do so.

Secondly - unfortunately I think the rewrite of gopkg.in -> github.com paths is necessary. The problem is that I removed the package `github.com/confluentinc/confluent-kafka-go/kafka/librdkafka` by renaming the `librdkafka` directory to `librdkafka_vendor`. However, `gopkg.in/confluentinc/confluent-kafka-go.v1/kafka` still depends on `github.com/confluentinc/confluent-kafka-go/kafka/librdkafka` (at least until this change is merged). So, the examples won't build on this branch without having done this.

I think our options are:
* Keep the change to rewrite the paths away from gopkg.in
* Back out the rewrite and just merge it with the failing example build
* Merge https://github.com/confluentinc/confluent-kafka-go/pull/553 first which should make this a non-issue

edenhill commented:
Why is there an import of librdkafka v1.5.2 in this PR? We're already on v1.5.2.


We'll stick with gopkg.in for a bit longer and then eventually switch over to modules only, so for now it's okay with failed example builds.

KJTsanaktsidis commented:
OK - I've backed out the import path changes.

> Why is there an import of librdkafka v1.5.2 in this PR?

I ran the import script to regenerate all the `build_*.go` files (so we can be sure that a) the import script works and b) that there won't be a pointless diff next time the script is run). I rebased the import commits away though so you can't tell anymore.

edenhill commented on kafka/librdkafka_vendor/import.sh:
```diff
@@ -48,7 +48,7 @@ if [[ -z $version ]]; then
 
 # Verify branch state
-curr_branch=$(git symbolic-ref HEAD 2>/dev/null | cut -d"/" -f 3)
+curr_branch=$(git symbolic-ref HEAD 2>/dev/null | cut -d"/" -f 3-)
```
Why this change? Can there be slashes in branches?

KJTsanaktsidis replied:
Yup, there definitely can - in fact this branch name has a slash in it too (`ktsanaktsidis/fix_dynamic_build`). At Zendesk it's conventional to structure branch names as `<your name>/<issue number>/<branch description>`, so I did this out of habit and found that the import script couldn't handle it.

edenhill reviewed (approved):
Great stuff! Just one comment

edenhill commented on kafka/kafka.go:
```diff
@@ -254,14 +254,14 @@ package kafka
 import (
 	"fmt"
 	// Make sure librdkafka/ sub-directory is included in vendor pulls.
```
librdkafka_vendor/

edenhill commented:
Are all comments addressed?

KJTsanaktsidis commented:
@edenhill Apologies - this totally fell of my radar! I've addressed the remaining feedback and I think this should be good to merge.

The PR is red because the examples won't build until it's merged as we discussed.

edenhill commented:
Thanks alot for this, @KJTsanaktsidis !

edenhill commented:
Huhm, I wonder where this import is coming from:
https://travis-ci.org/github/confluentinc/confluent-kafka-go/jobs/753330232#L615

KJTsanaktsidis commented:
Hum.... not some caching or something from gopkg.in is it? There's no release process or anything to make it start returning the new package right?

KJTsanaktsidis commented:
and `PKG_CONFIG_PATH=/usr/local/Cellar/openssl@1.1/1.1.1i/lib/pkgconfig go get -tags=dynamic ./...` works on my machine too :/

edenhill commented:
Maybe it's time to get rid of the gopkg.in imports..
